### DISCUSSION TITLE
Implementing manual (Class) based coroutines with .h and .cpp files

### DISCUSSION THREAD

**Discussion by Code-Assembly** [bxparks/AceRoutine#40] [General]
Hi, 

Right off the bat I would like to sincerely thank you for the awesome work you've done and this and your other libraries. Both the code quality and effort invested is truly commendable, and has certainly not gone unnoticed or unappreciated.

I'd like to ask if there is is any way in which the **Sound Manager** example could be implemented in a separate **.h** and **.cpp** file.
I would continuously get the following error ```undefined reference to `vtable for SoundManager'```

As such I eventually settled for implementing everything in a single separate **SoundManager.h file** which compiles without errors.

**Comment by bxparks** [comment]
Thanks for the encouraging words!

With regards to the error message, I have not seen this. Can you send information on how to reproduce? What board? What tool chain?

**Comment by Code-Assembly** [comment]
Hi silly of me not to include details:

Platform is **STM32Duino** on **Platform IO** the board is a **RobotDyn APM32F103CB** ( Effectively an STM32F103CB clone with 96MHz 128k Flash and 20k RAM )

I'm sure the issue is in my implementation so I'm rather hoping we can get an example that has a separate files but here is my non working implementation

**SoundManager.h**
```C++ 
#ifndef SoundManager_H
#define SoundManager_H

#include <Arduino.h>
#include <AceRoutine.h>

using namespace ace_routine;

enum Sound
{
    MUTE,
    NOTIFICATION,
    ALERT,
    WARNING,
    DANGER,
};

class SoundManager : public Coroutine
{
private:
    uint8_t _audioOutputPin;
    Sound _currentSound = MUTE;
    bool _loop;

public:
    void begin(uint8_t audioOutputPin);
    int virtual runCoroutine();
    void play(Sound sound, bool loop = false);
    void mute();
};

#endif
```

**SoundManager.cpp**
```C++
#include <Arduino.h>
#include "SoundManager.h"
#include "AudioNotes.h"

void SoundManager::begin(uint8_t audioOutputPin)
{
    _audioOutputPin = audioOutputPin;
}

int SoundManager::runCoroutine()
{
    COROUTINE_LOOP()
    {
        switch (_currentSound)
        {
        case Sound::NOTIFICATION:
            tone(_audioOutputPin, NOTE_B4);
            COROUTINE_DELAY(500);
            noTone(_audioOutputPin);
            COROUTINE_DELAY(125);

            tone(_audioOutputPin, NOTE_D4);
            COROUTINE_DELAY(125);
            noTone(_audioOutputPin);
            COROUTINE_DELAY(125 * 1.05);
            break;
        case Sound::ALERT:
            break;
        case Sound::WARNING:
            break;
        case Sound::DANGER:
            break;
        case Sound::MUTE:
            mute();
            break;
        }

        if (!_loop)
            mute();
    }
}

void SoundManager::play(Sound sound, bool loop)
{
    _currentSound = sound;
    _loop = loop;
}

void SoundManager::mute()
{
    _currentSound = Sound::MUTE;
    noTone(_audioOutputPin);
}
```
**Main.cpp**
```C++
#include <Arduino.h>
#include <AceCommon.h>
#include <AceRoutine.h>

#include "../lib/WLCConfig.h"
#include "../lib/SoundManager.h"

SoundManager soundManager;

void setup()
{
    pinMode(WLC::Pins::AUDIO_PWM_OUT, OUTPUT);
    soundManager.begin(WLC::Pins::AUDIO_PWM_OUT);
    soundManager.play(Sound::NOTIFICATION, true);
}

void loop()
{
    soundManager.runCoroutine();
}

```

**Comment by Code-Assembly** [comment]
Working implementation everything in one file

**SoundManager.h**
```C++
#ifndef SoundManager_H
#define SoundManager_H

#include <Arduino.h>
#include <AceRoutine.h>
#include "AudioNotes.h"

using namespace ace_routine;

enum Sound
{
    MUTE,
    NOTIFICATION,
    ALERT,
    WARNING,
    DANGER,
};

class SoundManager : public Coroutine
{
private:
    uint8_t _audioOutputPin;
    Sound _currentSound = MUTE;
    bool _loop;

public:
    void begin(uint8_t audioOutputPin)
    {
        _audioOutputPin = audioOutputPin;
    }

    int runCoroutine() override
    {
        COROUTINE_LOOP()
        {
            switch (_currentSound)
            {
            case Sound::NOTIFICATION:
                tone(_audioOutputPin, NOTE_B4);
                COROUTINE_DELAY(500);
                noTone(_audioOutputPin);
                COROUTINE_DELAY(125);
                
                tone(_audioOutputPin, NOTE_D4);
                COROUTINE_DELAY(125);
                noTone(_audioOutputPin);
                COROUTINE_DELAY(125 * 1.05);

                // COROUTINE_DELAY(1000);
                break;
            case Sound::ALERT:
                break;
            case Sound::WARNING:
                break;
            case Sound::DANGER:
                break;
            case Sound::MUTE:
                mute();
                break;
            }

            if (!_loop)
                mute();
        }
    }

    void play(Sound sound, bool loop = false)
    {
        _currentSound = sound;
        _loop = loop;
    }

    void mute()
    {
        _currentSound = Sound::MUTE;
        noTone(_audioOutputPin);
    }
};

#endif
```

**Comment by bxparks** [comment]
I don't use PlatformIO, so I can't provide official support for it. But can you paste in the *exact* error message from the compiler? Maybe I can see something obvious.

Also, you have tried compiling your program with the Arduino IDE or CLI? I have used AceRoutine coroutines defined in separate *.h and *.cpp files without problems in many personal projects.

**Comment by Code-Assembly** [comment]
The error i receive is

```shell
Building in release mode
Linking .pio/build/LaunchController/firmware.elf
/Users/aadil.cachalia/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: .pio/build/LaunchController/src/main.cpp.o: in function `_GLOBAL__sub_I_soundManager':
main.cpp:(.text.startup._GLOBAL__sub_I_soundManager+0x24): undefined reference to `vtable for SoundManager'
collect2: error: ld returned 1 exit status
*** [.pio/build/LaunchController/firmware.elf] Error 1
```

>  have used AceRoutine coroutines defined in separate *.h and *.cpp files without problems in many personal projects.

I'd really appreciate it I could see an example of how you do it. My limitation is likely in my lack of **c++** knowledge, of file linking and language syntax ect. that is the root of the issue.

**Comment by bxparks** [comment]
I don't know the PlatformIO environment, but the `main.cpp.o` seems strange to me. Usually the object file is named `main.o`. Did you try compiling from a clean, i.e. wiping out `*.o` files and any compiler cache?

I looked at my projects where I defined coroutines in separate `*.h` and `*.cpp` files, but unfortunately almost all of them are in private repos. The one example that is in a public repo is waaay too complicated to use as an example.

So I just submitted 40a5d55 where I extracted the `SoundManagerRoutine` and `SoundRoutine` in the `examples/SoundManager` directory into separate files. Take a look in my `develop` branch. I verified that this works under both Arduino IDE, Arduino CLI and under Linux (using EpoxyDuino). It ought work under PlatformIO if you configure it properly.

**Comment by bxparks** [comment]
(Transferring this to Discussions because I don't think it's a bug in the library.)

**Comment by Code-Assembly** [comment]
Thank you again for taking the time to do this. 
I'll have a look at the updated examples and post my an update to this thread should I succeed. :)

**Comment by Code-Assembly** [comment]
So the issue seems to have been that **extern** was the missing keyword


**main.cpp** : I updated the _soundManager_ variable declaration as follows
```cpp
extern SoundManager soundManager;
```

**SoundManager.h**: I changed my class declaration to include the _ace_routine_ namespace
```cpp
class SoundManager : public ace_routine::Coroutine
```

Thank you for helping me resolve this.

**Reply by bxparks** [reply]
So I considered a missing `extern` as the source of your problem, but that makes sense only if your were creating an instance of `SoundManager soundManager` in the `SoundManager.cpp` file, which you weren't. At least not in the version of the file that you posted. There should be only a single version of `SoundManager soundManager` without an `extern` and all other versions must have an `extern`.

With regards to the explicit namespace identifier in `public ace_routine::Coroutine`, that's a consequence of a good habit of never using a wildcard `using namespace xxx` statement in a header file. Because header files become included in other files, and the `using namespace xxx` will apply to every file where that is included, which pollutes the global namespace. So using a `using namespace xxx` is ok in a cpp file, but usually should be avoided in a header file.

Glad you got your code working.
